### DISCUSSION TITLE
Is the ESP32-PICO-D4 supported?

### DISCUSSION THREAD

**Discussion by bigdingoTECH** [someweisguy/esp_dmx#110] [General]
Hello, 

I am unsure if the preferred use of this forum is to post all your questions in one discussion but here goes...

Is the ESP32-PICO-D4 supported?

I am getting what appears to be a WDT reset loop printing to the serial monitor...

'11:20:30.869 -> rst:0x7 (TG0WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
11:20:30.869 -> flash read err, 1000
11:20:30.869 -> ets_main.c 371 
11:20:31.179 -> ets Jun  8 2016 00:22:57`

Additionally, I only require DMX transmit and don't require the EN pin as this is set in hardware, or the Receive pin, what should I declare these as?

Thank you so much for building this library, I am most grateful!

M

**Comment by someweisguy** [comment]
Thanks for reaching out! The PICO-D4 is supported as far as I understand. It appears to me that the PICO is just a standard ESP32 in a slightly different package. 

If you're getting watchdog errors, my best guess is that this library is clobbering your watchdog timer. If you are using Arduino or ESP-IDF v4.4, try using a different DMX_NUM - this will instruct the DMX library to use a different hardware timer which will hopefully make the error go away. 

If you aren't using the EN pin, you can use the constant `DMX_PIN_NO_CHANGE` and the EN pin will not be used.

Let me know if you have any other questions!

**Comment by bigdingoTECH** [comment]
Hello, @someweisguy.

So nice of you to send a kind and helpful message, I spend a lot of time on forums trying to work things out and the amount of vitriol gets to you after a while, your message was a refreshing change.

I have been looking into this further, your library works great on UTXD1 but caused the WDT reset loop on UTXD2.

Pin IO17 is the pin that the datasheet states is for UTXD2

`IO17 27 I/O GPIO17, HS1_DATA5, U2TXD, EMAC_CLK_OUT_180`

But also have just spotted this in the notes below.

`2. Pins IO16, IO17, CMD, CLK, SD0 and SD1 are used to connect the embedded flash, and can not be used for
other purposes. For details, please see Section 6 Schematics.
`
I think that in your DMX write example the default for 'int transmitPin = 17;' So perhaps it will help others if they find this post and are having similar issues.

How are you even supposed to use UTXD2 if it is connected to embedded flash?

I now have the issue that I wanted to use UTXD1 for something else! Any ideas, how to get two working UARTS?

Thank you! again!

Best, 

Matt

**Reply by someweisguy** [reply]
> I think that in your DMX write example the default for 'int transmitPin = 17;' So perhaps it will help others if they find this post and are having similar issues.

Good spotting! This is helpful. I used to have all the examples use UART 2 but recently changed the examples to UART 1. I must have forgotten to update the pin numbers. I will do so soon!

I looked at the PICO datasheet [here](https://www.espressif.com/sites/default/files/documentation/esp32-pico-d4_datasheet_en.pdf) and it does indeed appear that GPIO17 is hardwired directly to embedded flash, so it is unusable. Perhaps you can use UART 2 by attaching it to a different pin? For example, you can try attaching UART 2 to GPIO32 and GPIO33. This won't fix the WDT errors if you are still having them.

**Reply by bigdingoTECH** [reply]
Thank you @someweisguy,

is the UART being used chosen by...?

`dmx_port_t dmxPort = 1;`

And the pins used chosen by...?

`int transmitPin = X;
int receivePin = X;`

So you could expect something like this to work?

`int transmitPin = 32;
int receivePin = DMX_PIN_NO_CHANGE;
int enablePin = DMX_PIN_NO_CHANGE;
dmx_port_t dmxPort = 2;
byte data[DMX_PACKET_SIZE];`

Do I need to do anything else to reprovision Uart pins? Use the HardwareSerial library or something?

Best,

Matt

**Reply by someweisguy** [reply]
You are correct! Set the `dmxPort` to the appropriate DMX port you would like to use. To set the pins, you pick which pins you would like to use and call `dmx_set_pin()`. The set_pin function will provision the pins to the desired DMX port. It will also print an error to the console if the pins you have chosen are not able to be used by your ESP32.

You do not need to call any other functions to reprovision the UART pins. :)
